### DISCUSSION TITLE
Enforce scope properly for attachment and other tables

### DISCUSSION THREAD

**Discussion by grantfitzsimmons** [specify/specify7#5730] [Feature Requests]
https://github.com/specify/specify6/blame/8c853f91b76ff7633c8318aff5008e1822ca10a0/src/edu/ku/brc/specify/datamodel/Attachment.java

This is a mess, `collectionmemberid` integer fields are not the way to go

**Comment by maxpatiiuk** [comment]
Can confirm it's a giant mess.
TLDR: leave it be the way it is, and when sp7 is no longer inprisoned by sp6, re-engineer the scoping from scratch.

Findings:
Some tables have a direct relationship to scoping tables (CollectionObject, Collection, Discipline, Division, Institution). Scoping for those is enforced in sp6. Scoping for them is enforced in sp7 only in query results and when going to that record directly by the URL - not enforced in other places.
There is an exception for query builder. The query builder scopes to Division only for some tables. See https://github.com/specify/specify7/commit/f79e70cf19c07354f7e1f8ee5167e82fcf9b27ae

All of the below scoping is enforced in sp6 only using some custom logic for every case. None of this is implemented in sp7 (unless mentioned otherwise):
1. Children tables inherit scope from the parent (Accession Authorization is scoped to the same level as Accession)
2. Attachments are scoped using `scopeType` ([represents id of the](https://github.com/specify/specify6/blame/8c853f91b76ff7633c8318aff5008e1822ca10a0/src/edu/ku/brc/specify/datamodel/Attachment.java#L74-L78)) and `scopeId` fields (represents the ID of the record in that scope). Note, the `scopeId` may refer to the record that no longer exists as delete blockers don't check for those (since `scopeId` is just a numeric field.
3. `SpPermission` table stores sp6 permissions. It has it's own weird solution for scoping.
4. There is an `SpPrincipal` table. `RecordSet`, `AppResource`, `SpPermission`, `Workbench` tables have relationship to it. The table appears to be used for permissions or scoping or both.
5. The following tables have a `collectionMemberId` **integer** field:
    ```json
    [
      "Borrow",
      "BorrowAgent",
      "BorrowMaterial",
      "BorrowReturnMaterial",
      "CollectingEventAttachment",
      "CollectingEventAttr",
      "CollectingTripAttachment",
      "CollectionObject",
      "CollectionObjectAttachment",
      "CollectionObjectAttr",
      "CollectionObjectAttribute",
      "CollectionObjectCitation",
      "CollectionObjectProperty",
      "Container",
      "DNASequence",
      "DNASequencingRun",
      "Determination",
      "DeterminationCitation",
      "InfoRequest",
      "MaterialSample",
      "OtherIdentifier",
      "Preparation",
      "PreparationAttachment",
      "PreparationAttr",
      "PreparationAttribute",
      "PreparationProperty",
      "Project",
      "RecordSet",
      "SpExportSchemaMapping",
      "SpFieldValueDefault",
      "SpSymbiotaInstance",
      "VoucherRelationship"
    ]
    ```
   
    From what we can tell, that was a scoping solution inherited from Specify 5. The `collectionMemberId` does not consistency constraints so it may refer to collections that no longer exist.

    Specify 7 appears to be checking this field in tree stats and query builder scoping, but only for some tables.


Impact:
Scoping is not enforced consistently between tables, let alone between sp6 and sp7.
Sp6 has way too many messy ways to do scoping.
The best solution would be to leave things be until sp6 support is dropped, at which point there can be a database migration that fixes scoping for all tables.

##### Proposed ideal solutions for scoping:

#### Simple
If need to support dynamic scoping (scoping to different levels for different relationships), do what SpAppResourceDir table does and have optional relationships to `collection` or `division`. Then, if collection relationship is not null, then it is scoped to collection. Else, If division is not null, then it is scoped to division. Else, it is not scoped.

If table is always scoped to the same level, then just have a required relationship to `collection` or whatever it is.

#### Advanced
Have a `Scope` table, where each row could have an optional relationship to all scoping tables as separate columns.
Then, each record in the database that needs to be scoped, just has a relationship to the Scope record that contains the desired scope (has relationship to desired collection/discipline/division).

Then, imagine an advanced use case: when running a query, and for example want to scope results to two queries (e.x, because user said they don't want scoring, but they only have access to two collections), you just get ID of all the records in the `Scope` table that match your desired scope, and then filter all query results to records where `scopeId` relationship leads to one of the Scoes that matches the desired scope.

This solution can be easily extended in the future ot add more scoping by just adding new relationship to the `Scope` table.

What is more, the `Scope` table solution can be texted to accommodate customizing the scoping hierarchy (https://github.com/specify/specify7/issues/77)

**Comment by maxpatiiuk** [comment]
Update:
A list of resolved scoping paths that sp7 front-end recognizes - https://github.com/specify/specify7/blob/368eddf92b5d5c3fc9583cb20132d5f8cda3ec93/specifyweb/frontend/js_src/lib/components/DataModel/__tests__/__snapshots__/specifyModel.test.ts.snap#L1029-L1217

Only direct relationships to scoping table are currently enforced in forms and query builder (i.e, Accession: 'division').
This is to preserve the compatibility with the previous behavior.
We can consider changing that in the future.

Back-end query builder also looks at collectionMemberId field. Front-end does not. All other non-relationship scoping fields are ignored by both front-end and back-end.

**Comment by maxpatiiuk** [comment]
Okay, it get's worse. Turns out collection object has both collection relationship and collection id field

query builder bakc-end to look at collection relationship first (in this case, can't guarantee all)

front-end happens to look at collection id field (in this case, can't guarantee all)

![Screenshot 2023-04-14 at 11 47 09](https://user-images.githubusercontent.com/40512816/232106649-87a82b00-4950-4d5a-ac30-e75444f2025a.png)

**Comment by melton-jason** [comment]
> Okay, it get's worse. Turns out collection object has both collection relationship and collection id field
> 
> query builder bakc-end to look at collection relationship first (in this case, can't guarantee all)
> 
> front-end happens to look at collection id field (in this case, can't guarantee all)
> 
> ![Screenshot 2023-04-14 at 11 47 09](https://user-images.githubusercontent.com/40512816/232106649-87a82b00-4950-4d5a-ac30-e75444f2025a.png)

There are 32 tables which have 'CollectionMemberID' column. Preferably, we would be consistent in the way we manage scoping like this, but it will have to wait until we break compatibility with Specify 6. 

For reference, the 32 tables are:

```sql
SELECT DISTINCT table_name FROM information_schema.columns WHERE table_schema="specify" AND column_name="collectionmemberid";
```

+----------------------------+
| preparationattr           |
| borrowagent                |
| project                    |
| otheridentifier            |
| container                  |
| dnasequence                |
| collectingtripattachment   |
| borrowreturnmaterial       |
| materialsample             |
| preparation                |
| collectionobjectattachment |
| borrow                     |
| spfieldvaluedefault        |
| collectionobjectproperty   |
| collectingeventattachment  |
| dnasequencingrun           |
| spsymbiotainstance         |
| preparationattachment      |
| voucherrelationship        |
| collectingeventattr        |
| spexportschemamapping      |
| collectionobject           |
| determinationcitation      |
| preparationproperty        |
| collectionobjectattribute  |
| determination              |
| recordset                  |
| collectionobjectattr       |
| collectionobjectcitation   |
| inforequest                |
| borrowmaterial             |
| preparationattribute       |
+----------------------------+
